distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerCli.java

```diff
@@ -29,21 +29,30 @@
 
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.List;
 import java.util.Locale;
 
 /**
  * The main CLI for running Elasticsearch.
  */
 class ServerCli extends EnvironmentAwareCommand {
 
+    interface ServerProcessPostExitProcessor {
+        void process(int exitCode);
+
+        void waitForCompletion() throws InterruptedException;
```

Comment:
A different name here would be more appropriate - `processExiting()` as a handler or listener-type method, `waitForCompletion` implies `ServerCli` is waiting for the processor to actively do something before it can exit.

Reply:
I have an idea that I think I like better:
```
interface Awaitable {
    void await();
}

interface ServerProcessPostExitProcessor {
   Awaitable startProcessing(int exitCode);
}

int execute() {
   ...
   for (final var postExitProcessor : serverProcessPostExitProcessors) {
      awaitables.add(postExitProcessor.startProcessing(exitCode));
   }
}

void close() {
    ...
    foreach (var awaitable: awaitables) {
       awaitable.await();
    }
}
```
It conveys more clearly that it is a async operation and that the caller needs to/can await on it

Reply:
No, while that looks more elegant, it would make the time relation between process exit and start of processing less clear. Better to encapsulate things in a abstract class that implements the await (so things can be awaited even before processing started) and leave the user just the `process` part to implement.
I'll submit another revision that makes the mechanism more clear and less error prone.